File: rust/src/phase_correct.rs
//! Phase-ambiguity correction (the complex-domain analogue of real ICA's sign fix).
//!
//! Complex ICA recovers each source only up to a complex scalar c_k * e^{i theta_k}, so
//! the split of a component's energy between real and imaginary parts is arbitrary.
//! Convention: rotate each component so its energy is maximally concentrated in the real
//! part. The voxel values form an elongated cloud ~ r(v) e^{i alpha}; squaring maps the
//! +/-alpha line ambiguity to the single angle 2*alpha, giving the closed form
//! theta = -0.5 * arg(sum s^2).
//!
//! Reference: Rodriguez, Calhoun & Adali (2012), Pattern Recognition 45:2050-2063.

use nalgebra::DMatrix;
use num_complex::Complex64;

fn skewness(x: &[f64]) -> f64 {
    let n = x.len() as f64;
    let mean = x.iter().sum::<f64>() / n;
    x.iter().map(|v| (v - mean).powi(3)).sum::<f64>() / n
}

/// Rotate each component onto the real axis; apply the inverse rotation to `a` so that
/// `a * s` is preserved exactly. Returns the applied `theta` per component.
pub fn correct_phase(
    s: &mut DMatrix<Complex64>,
    a: &mut DMatrix<Complex64>,
    mask: Option<&[bool]>,
) -> Vec<f64> {
    let (n, v) = s.shape();
    let keep: Vec<usize> = match mask {
        Some(m) => (0..v).filter(|&i| m[i]).collect(),
        None => (0..v).collect(),
    };

    let mut theta = vec![0.0; n];
    for k in 0..n {
        // theta = -0.5 * arg(sum of squares over masked voxels)
        let sum_sq: Complex64 = keep.iter().map(|&i| s[(k, i)] * s[(k, i)]).sum();
        let mut th = -0.5 * sum_sq.arg();

        let rot = Complex64::new(th.cos(), th.sin());
        for i in 0..v {
            s[(k, i)] *= rot;
        }

        // residual pi ambiguity: make the real-part skewness positive.
        // Fold the flip into theta BEFORE rotating `a`, or the reconstruction breaks.
        let re: Vec<f64> = keep.iter().map(|&i| s[(k, i)].re).collect();
        if skewness(&re) < 0.0 {
            th += std::f64::consts::PI;
            for i in 0..v {
                s[(k, i)] = -s[(k, i)];
            }
        }

        let inv = Complex64::new(th.cos(), -th.sin()); // e^{-i th}
        for r in 0..a.nrows() {
            a[(r, k)] *= inv;
        }
        theta[k] = th;
    }
    theta
}

/// Align each subject component to a reference (e.g. the group/aggregate map).
///
/// Per-subject phase corrections are independent, so without this the residual rotations
/// reintroduce non-physiological variance before group statistics.
pub fn align_to_reference(
    s: &mut DMatrix<Complex64>,
    s_ref: &DMatrix<Complex64>,
) -> Vec<f64> {
    let (n, v) = s.shape();
    let mut theta = vec![0.0; n];
    for k in 0..n {
        let dot: Complex64 = (0..v).map(|i| s_ref[(k, i)].conj() * s[(k, i)]).sum();
        let th = -dot.arg();
        let rot = Complex64::new(th.cos(), th.sin());
        for i in 0..v {
            s[(k, i)] *= rot;
        }
        theta[k] = th;
    }
    theta
}

File: python/src/gift/phase_correct.py
"""Phase-ambiguity correction (the complex-domain analogue of real ICA's sign fix).

Complex ICA recovers each source only up to a complex scalar c_k * exp(i*theta_k):

    X = A @ S  <=>  X = (A @ inv(D)) @ (D @ S),   D = diag(c_k * exp(i*theta_k))

so the split of a component's energy between real and imaginary parts is arbitrary.
Convention: rotate each component so its energy is maximally concentrated in the real
part. The voxel values form an elongated cloud ~ r(v) * exp(i*alpha); squaring maps the
+/-alpha line ambiguity to the single angle 2*alpha, giving the closed form
theta = -0.5 * angle(sum(s**2)).

Reference: Rodriguez, Calhoun & Adali (2012), Pattern Recognition 45:2050-2063.
"""

import numpy as np
from scipy.stats import skew


def correct_phase(S, A, mask=None):
    """Rotate each component onto the real axis; apply the inverse to A.

    S: (N, V) complex sources. A: (M, N) complex mixing. Returns (S, A, theta).
    A @ S is preserved exactly.
    """
    S = np.array(S, dtype=np.complex128, copy=True)
    A = np.array(A, dtype=np.complex128, copy=True)
    N, V = S.shape
    if mask is None:
        mask = np.ones(V, dtype=bool)
    mask = np.asarray(mask, dtype=bool).ravel()

    theta = np.zeros(N, dtype=np.float64)
    for k in range(N):
        sm = S[k, mask]
        th = -0.5 * np.angle(np.sum(sm**2))  # orient the major axis to the real axis
        sk = S[k] * np.exp(1j * th)
        # residual pi ambiguity: fix the direction by real-part skewness
        if skew(sk[mask].real) < 0:
            th += np.pi
            sk = -sk
        S[k] = sk
        A[:, k] = A[:, k] * np.exp(-1j * th)
        theta[k] = th
    return S, A, theta


def align_to_reference(S, S_ref):
    """Align each subject component to a reference (e.g. the group/aggregate map).

    Per-subject phase corrections are independent, so without this the residual
    rotations reintroduce non-physiological variance before group statistics.
    Returns (S_aligned, theta).
    """
    S = np.array(S, dtype=np.complex128, copy=True)
    S_ref = np.asarray(S_ref, dtype=np.complex128)
    N = S.shape[0]
    theta = np.zeros(N, dtype=np.float64)
    for k in range(N):
        th = -np.angle(np.sum(np.conj(S_ref[k]) * S[k]))
        S[k] = S[k] * np.exp(1j * th)
        theta[k] = th
    return S, theta
